feat: INFRA-132 Fixed update user secret

ihannache opened the pull request:
I have an existing minio user, when I try to update the secret for this user it's not working. the user still have the old password.

This PR implements the following changes:
```
func minioUpdateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {

	iamUserConfig := IAMUserConfig(d, meta)

	var err error
	secretKey := iamUserConfig.MinioSecret

	if secretKey == "" || iamUserConfig.MinioUpdateKey {
		if secretKey, err = generateSecretAccessKey(); err != nil {
			return NewResourceError("error creating user", d.Id(), err)
		}
	}
the fix :

func minioUpdateUser(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {

	iamUserConfig := IAMUserConfig(d, meta)

	var err error
	secretKey := iamUserConfig.MinioSecret

---> if secretKey == "" {
		if secretKey, err = generateSecretAccessKey(); err != nil {
			return NewResourceError("error creating user", d.Id(), err)
		}
	}
```
## Reference

 - Resolves: #354

## Closing issues

- Closes: #354

Magnitus- commented on minio/resource_minio_iam_user.go:
```diff
@@ -146,7 +155,7 @@ func minioUpdateUser(ctx context.Context, d *schema.ResourceData, meta interface
 	}
 
-	if iamUserConfig.MinioUpdateKey {
+	if iamUserConfig.MinioUpdateKey || d.HasChangeExcept(iamUserConfig.MinioSecret){
```
What you want to put there, is actually `d.HasChange("secret")`, not `d.HasChangeExcept(iamUserConfig.MinioSecret)`

Magnitus- commented on minio/resource_minio_iam_user.go:
```diff
@@ -124,10 +125,18 @@ func minioUpdateUser(ctx context.Context, d *schema.ResourceData, meta interface
+			AccessKey: iamUserConfig.MinioIAMName,
+			SecretKey: secretKey,
+			Status:    madmin.AccountEnabled,
+		}	
```
This looks like it would break the secret generation/auto-update functionality (anyways, my understanding of it), except that it didn't seem to work anyhow with the validation I did locally so I really don't know.

BuJo commented:
If you have the time, could you add a test for this problem for verification?  I tried to analyze this in #354 but couldn't reproduce it with the setup we have.  Some things behave slightly differently depending on the version of Terraform used to run the provider, perhaps its that, or..

Magnitus- commented:
> If you have the time, could you add a test for this problem for verification? I tried to analyze this in #354 but couldn't reproduce it with the setup we have. Some things behave slightly differently depending on the version of Terraform used to run the provider, perhaps its that, or..

I validated on Terraform 1.2.9 on a local dockerized minio version 2022-10-15T19-57-03Z.

I validated:
- The latest version of this provider at the time (1.7.1 I believe)
- Issam's branch
- Our fork with my further additions

For the first two, I validated with update_secret at true and false.

The changing of the password by changing the secret input didn't work in the latest version of your provider. Issam's changed fixed that, but in a way that felt like a side-effect of the code change (hence my comment). 

Also, his change seemed to break the 'update_secret' feature (when I ran it in my head), but empirically, the feature didn't seem to work anyways (not according ot my understanding of it anyhow).

My understanding of the 'update_secret' flag is that if true, the provider will auto-update the secret with a generated internal value on each terraform apply, but I never could get that to work (I tried it with passing an initial secret and not passing an initial secret) and frankly, I found that the provider supporting the generation and storage of the user password complexified the code of the provider a lot: You essentially have to handle re-creation on user name change in the updater instead of letting 'force new' do the work for you (in order not to lose the generated password in the provider) and that seems to break the re-creation of the user on name change btw.

For the complexity is adds, I personally don't consider the feature that much of a gain (though you might have a specific use-case we don't here) since you can simply create a separate random password resource separately outside this provider if you want to manage that workflow in terraform or otherwise use some other provider to fetch the password from an external store like vault.

BuJo commented:
Trying again to wrap my head around this and analyzing the respective test (`TestAccAWSUser_RotateAccessKey`).

```tf
resource "minio_iam_user" "test3" {
  name          = "test-user-4857382394561118561"
}
```
to
```tf
resource "minio_iam_user" "test3" {
  update_secret = true
  name          = "test-user-4857382394561118561"
}
```

Generates the password and then generates a different password.  I manually checked via breakpoints that the user actually was changed in the minio instance.

I installed the provider locally and made the same test using the same scheme.  First a user, then the same one with `update_secret`.  Same result.

* Provider version: `1.8.0` (also tried with `1.7.1`)
* Terraform version: `1.3.3` (also tried with `1.2.9` manually downloaded)

```tf
terraform {
  required_providers {
    minio = {
      source = "aminueza/minio"
      version = ">= 1.8.0"
    }
  }
  required_version = ">= 1.3"
}

provider "minio" {
  minio_server   = "localhost:9000"
  minio_region   = "us-east-1"
  minio_user     = "minio"
  minio_password = "minio123"
}

resource "minio_iam_user" "test3" {
  name          = "test-user-jo"
#  update_secret = true
}
```

Are you doing something different?

Answering my own question: yes.  Meant is the manual changing of secrets, not the automatic way.

I verified that behaviour and will push a patch.

ihannache commented:
Hello @BuJo,

I added your patch to my branch, the good news is that the secret update works fine, I changed the secret field  many times and I can login with it.
The bad news is that when I try to update the user's name, the apply fails, here is the details :

```
terraform apply the next ressource: 

resource "minio_iam_user" "test" {
   name = "test"
   secret = "test1234"
}

 then,
change the name from test to test-123:

issamhannache@MacBook-Pro terraform % terraform apply
minio_iam_user.test: Refreshing state... [id=test]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # minio_iam_user.test will be updated in-place
  ~ resource "minio_iam_user" "test" {
        id            = "test"
      ~ name          = "test" -> "test-123"
      ~ secret        = (sensitive value)
        # (4 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

minio_iam_user.test: Modifying... [id=test]
╷
│ Error: [FATAL] error to disable IAM User %s: %s (test): The specified user does not exist. (Specified user does not exist)
│ 
│   with minio_iam_user.test,
│   on users.tf line 5, in resource "minio_iam_user" "test":
│    5: resource "minio_iam_user" "test" {
│ 
```

BuJo reviewed (approved):
Thank you for testing as well @Magnitus- !  Allright, let's merge this one, and handle the renaming of users in a separate commit as I haven't put enough thought into that one.

BuJo commented:
@Magnitus- Could you `go fmt` the following files (Using go 1.19)?  I guess then it's good to go!

```
Error: The following files are not properly formatted:
Error:   minio/resource_minio_iam_user.go
Error:   minio/resource_minio_iam_user_test.go
```

BuJo commented:
Since I repaired the Login-Check, two more tests fail, sorry, I didn't run all tests after checking that the one was good...  I'll have a look at it.

Magnitus- commented:
> Thank you for testing as well @Magnitus- ! Allright, let's merge this one, and handle the renaming of users in a separate commit as I haven't put enough thought into that one.

At the risk of adding a breaking change here, I think it would be worthwhile to do a serious analysis about the need to manage a generated secret in the user resource. If you remove all the logic in the user resource related to that (and just put a "force new" on the user name), it works fine.

My intuition is that the more idiomatic way to do it in terraform (as suggested by their own random provider) would be to generate the secret separately like so:

```
resource "random_password" "user" {
  length           = 30
  lower            = true
  number           = true
  upper            = true
  special          = false
}

resource "minio_iam_user" "user" {
  name               = "user"
  update_secret = random_password.user.result
}
```

If you really wish to manage automatic changing of passwords (and perhaps customized password formats), you could create a separate credentials resource like so:

```
resource "minio_secret_key" "user" {
  update_secret = true
}

resource "minio_iam_user" "user" {
  name               = "user"
  update_secret = minio_secret_key.user.result
}
```

BuJo commented:
Yeah, it's breaking my brain a little bit, me wanting to retain backwards compatibility and fixing the behaviour at the same time.

My reasoning for using the simple `ForceNew` would be that if you can revise the username, then you can revise the password at the same time.  I'm unsure if that is satisfactory however.

Since #376 is through, the unrelated test failure won't be hitting this PR.  We should rebase this PR on top of the current master.

I've crafted a patch, which tries to clarify the code a little and finally satisfies all the tests.

```patch
From 6d4d04937295b4c766b47b0255c83148c7f6eb18 Mon Sep 17 00:00:00 2001
From: Jonathan Buch <jbuch@synyx.de>
Date: Thu, 27 Oct 2022 10:15:58 +0200
Subject: [PATCH] iam user, fix secret generation

---
 minio/resource_minio_iam_user.go | 37 ++++++++++++++------------------
 1 file changed, 16 insertions(+), 21 deletions(-)

diff --git a/minio/resource_minio_iam_user.go b/minio/resource_minio_iam_user.go
index ba0f4e2..9b69a8f 100644
--- a/minio/resource_minio_iam_user.go
+++ b/minio/resource_minio_iam_user.go
@@ -99,24 +99,9 @@ func minioUpdateUser(ctx context.Context, d *schema.ResourceData, meta interface
 
 	iamUserConfig := IAMUserConfig(d, meta)
 
-	var err error
-	secretKey := iamUserConfig.MinioSecret
-
-	if secretKey == "" || iamUserConfig.MinioUpdateKey {
-		if secretKey, err = generateSecretAccessKey(); err != nil {
-			return NewResourceError("error creating user", d.Id(), err)
-		}
-		_ = d.Set("secret", secretKey)
-	}
-
-	userStatus := UserStatus{
-		AccessKey: iamUserConfig.MinioIAMName,
-		SecretKey: secretKey,
-		Status:    madmin.AccountEnabled,
-	}
-
+	wantedStatus := madmin.AccountEnabled
 	if iamUserConfig.MinioDisableUser {
-		userStatus.Status = madmin.AccountDisabled
+		wantedStatus = madmin.AccountDisabled
 	}
 
 	if iamUserConfig.MinioForceDestroy {
@@ -124,18 +109,28 @@ func minioUpdateUser(ctx context.Context, d *schema.ResourceData, meta interface
 	}
 
 	userServerInfo, _ := iamUserConfig.MinioAdmin.GetUserInfo(ctx, iamUserConfig.MinioIAMName)
-	if userServerInfo.Status != userStatus.Status {
-		err := iamUserConfig.MinioAdmin.SetUserStatus(ctx, userStatus.AccessKey, userStatus.Status)
+	if userServerInfo.Status != wantedStatus {
+		err := iamUserConfig.MinioAdmin.SetUserStatus(ctx, iamUserConfig.MinioIAMName, wantedStatus)
 		if err != nil {
 			return NewResourceError("error to disable IAM User %s: %s", d.Id(), err)
 		}
 	}
 
-	if d.HasChange("secret") {
-		err := iamUserConfig.MinioAdmin.SetUser(ctx, userStatus.AccessKey, userStatus.SecretKey, userStatus.Status)
+	wantedSecret := iamUserConfig.MinioSecret
+	if iamUserConfig.MinioUpdateKey {
+		if secretKey, err := generateSecretAccessKey(); err != nil {
+			return NewResourceError("error creating user", d.Id(), err)
+		} else {
+			wantedSecret = secretKey
+		}
+	}
+
+	if d.HasChange("secret") || iamUserConfig.MinioSecret != wantedSecret {
+		err := iamUserConfig.MinioAdmin.SetUser(ctx, iamUserConfig.MinioIAMName, wantedSecret, wantedStatus)
 		if err != nil {
 			return NewResourceError("error updating IAM User Key %s: %s", d.Id(), err)
 		}
+		_ = d.Set("secret", wantedSecret)
 	}
 
 	return minioReadUser(ctx, d, meta)
-- 
2.37.3
```

Magnitus- commented:
> Are you doing something different?
> 
> Answering my own question: yes. Meant is the manual changing of secrets, not the automatic way.
> 
> I verified that behaviour and will push a patch.

My methodology was as follows...

First, I setup a minio server in one terraform workspace:

```
provider "kubernetes" {
  config_path    = "~/.kube/config"
  config_context = "minikube"
}

resource "kubernetes_deployment" "minio" {
  metadata {
    name = "minio"
    labels = {
      app = "minio"
    }
  }

  spec {
    replicas = 1

    selector {
      match_labels = {
        app = "minio"
      }
    }

    template {
      metadata {
        labels = {
          app = "minio"
        }
      }

      spec {
        container {
          image = "quay.io/minio/minio:RELEASE.2022-10-15T19-57-03Z"
          name  = "minio"
          command = ["minio", "server", "/data", "--console-address", "0.0.0.0:9001"]
        }
      }
    }
  }
}

resource "kubernetes_service" "minio" {
  metadata {
    name = "minio"
  }
  spec {
    selector = {
      app = "minio"
    }

    port {
      name        = "api"
      port        = 9000
      target_port = 9000
      node_port = 31900
    }

    port {
      name        = "console"
      port        = 9001
      target_port = 9001
      node_port = 31901
    }

    type = "NodePort"
  }
}
```

In another workspace:

```
provider minio {
  minio_server       = "127.0.0.1:31900"
  minio_access_key   = "minioadmin"
  minio_secret_key   = "minioadmin"
  minio_ssl          = false
}

resource "minio_iam_user" "test" {
   name = "test3"
   update_secret = true
}

resource "local_file" "test" {
    content  = minio_iam_user.test.secret
    filename = "${path.module}/test_password"
}
```

I ran terraform apply repeatedly and the secret didn't update in the file.

But I think I understand the meaning behind the flag for you now. This is not a flag that you meant to leave on to continuously update on each apply.

Instead, this is a flag that you add once to force an update and then remove afterwards and re-add the next time you want to force an update.

BuJo commented:
This has been merged in #425
